<a href="https://colab.research.google.com/github/joeyMckinney/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/Josiah_McKinney_LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [5]:
# YOUR WORK HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-07-14 22:47:18--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip.1’

student.zip.1       100%[===================>]  20.00K  --.-KB/s    in 0.1s    

2020-07-14 22:47:18 (154 KB/s) - ‘student.zip.1’ saved [20478/20478]



In [6]:
!unzip student.zip

Archive:  student.zip
replace student-mat.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: student-mat.csv         
replace student-por.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: student-por.csv         
replace student-merge.R? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: student-merge.R         
replace student.txt? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: student.txt             


In [12]:
import numpy as np
from scipy.stats import chisquare
import pandas as pd
from scipy import stats

In [8]:
df = pd.read_csv('student-mat.csv', sep=';')
print(df.shape)
df.head()

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

In [9]:
df.describe(exclude='number')

school  sex address famsize  ... nursery higher internet romantic
count     395  395     395     395  ...     395    395      395      395
unique      2    2       2       2  ...       2      2        2        2
top        GP    F       U     GT3  ...     yes    yes      yes       no
freq      349  208     307     281  ...     314    375      329      263

[4 rows x 17 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [11]:
# YOUR WORK HERE
contingency1 = pd.crosstab(df['failures'], df['internet'])
contingency1

internet  no  yes
failures         
0         47  265
1         13   37
2          2   15
3          4   12

1) null hypothesis: Independent

2) alternative hypothesis: Dependent

3) Confidence level: 95%

In [14]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency1)
print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 4.796825454131461
p value 0.18729362004759023
degrees of freedom 3
expected frequencies table 
 [[ 52.13164557 259.86835443]
 [  8.35443038  41.64556962]
 [  2.84050633  14.15949367]
 [  2.67341772  13.32658228]]


Conclusion:

due to a chi^2 statistic of 4.7968 and a p-value of .187, we fail to reject the null hypothesis that internet and failures are Independent.

In [27]:
contingency2 = pd.crosstab(df['freetime'], df['goout'])
contingency2

goout     1   2   3   4   5
freetime                   
1         4   8   5   1   1
2         5  27  17  11   4
3         5  41  70  26  15
4         5  18  34  41  17
5         4   9   4   7  16

1) null hypothesis: Independent

2) alternative hypothesis: Dependent

3) Confidence level: 95%

In [28]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency2)
print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 80.87785174188721
p value 1.155517403684047e-10
degrees of freedom 16
expected frequencies table 
 [[ 1.10632911  4.95443038  6.25316456  4.13670886  2.54936709]
 [ 3.72658228 16.68860759 21.06329114 13.93417722  8.58734177]
 [ 9.14177215 40.93924051 51.67088608 34.18227848 21.06582278]
 [ 6.69620253 29.98734177 37.84810127 25.03797468 15.43037975]
 [ 2.32911392 10.43037975 13.16455696  8.70886076  5.36708861]]


Conclusion:

due to a chi^2 statistic of 80.87 and a p-value of ~0, we reject the null hypothesis that freetime and goout are dependent.

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [ ]:
# YOUR WORK HERE

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)